docs/proposals/provisioning-proposal.md

```diff
@@ -0,0 +1,112 @@
+
+Cluster initialization and node joining are going to be configured using Kubeadm configuration files, which provide a big set of options for configuring cluster and all relevant cluster components (such as kube-controller-manager, kube-proxy, etc.). Cluster-API users would write kubeadm configuration files themselves and pass them as a ConfigMap/Secret or pass them to the function for generating userdata.
+
+Installing any other packages or manifests, including CNI, are not scope of those scripts, neither scope of Cluster-API. Cluster-API is not supposed to be a package manager, so everything needed should be installed as an add-on.
+
+### Implementation Prerequisites
+
+1. clusterctl should be able to install CNI. This was discussed on Cluster-API Breakout meeting held on 10/10. [Relevant issue #534](https://github.com/kubernetes-sigs/cluster-api/issues/534)
+
+### Implementation details
+
+The implementation will be based on [Go Templates](https://golang.org/pkg/text/template/). That allows implementers to source environment variables, pass data such as IP addresses, host names, and kubeadm tokens, configure Kubelet and Cloud Provider, without any change to the Cluster-API source code.
+
+Similar approach is used by many projects and providers, including:
+
+* [Cluster-API provider for DigitalOcean](https://github.com/kubermatic/cluster-api-provider-digitalocean/blob/master/cloud/digitalocean/actuators/machine/userdata.go)
+* [Cluster-API provider for GCP](https://github.com/kubernetes-sigs/cluster-api-provider-gcp/blob/master/pkg/cloud/google/metadata.go)
+* [Cluster-API provider for AWS](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/4553a80b6337b4adcc378c07db943772d30fbc78/pkg/cloud/aws/services/ec2/bastion.go)
+* [Cluster-API provider for VSphere](https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/master/cloud/vsphere/provisioner/common/templates.go)
+
+The implementation will be realized using interfaces, meaning that for each supported operating system interface will be implemented. The interface has two functions, for getting user data for Master instances and for workers.
+
+Both functions should take the following arguments:
+
+* Environment variables, provided as a string map and then parsed and appended on the top of script.
+* Kubernetes version, including Kubelet, kubeadm, kubernetes-cni versions.
+* Kubelet and Cloud Provider configuration files
+* To be discussed how we want to provide those:
+* Kubeadm configuration file. Do we want to be able to pass it clusterctl, as a Secret/ConfigMap, or as a function parameter?
+
+```go
+type Userdata interface {
+    MasterUserData(envVars map[string]string, kubeVersion string, cniVersion string) (string, error)
+    NodeUserData(envVars map[string]string, kubeVersion, cniVersion string) (string, error)
+}
+```
+
+### Decisions to be made
+
+* What Kubernetes versions should be supported?
+    * Kubernetes supports 3 version at the same time.
```

Comment:
@roberthbailey: I just reviewed https://github.com/kubernetes-sigs/cluster-api-provider-gcp/pull/54 where we updated bootstrap scripts to a different k8s version and the scripts had to be changed. 

And if we look at the history of kubeadm, I expect constant changes in the flags / configuration that we need to pass into kubeadm itself as well as the system components. We are constantly adding new flags / behavior to k8s that needs to be configured as we move to new releases. That needs to be reflected somewhere in the configuration (at least for the control plane components).